daemon: refactor Hubble Exporters as a cell

devodev opened the pull request:
Continuation of https://github.com/cilium/cilium/pull/35206

- Introduce a new `hubble-exporters` cell that provides the static and dynamic exporters using observer Options.
- The static exporter now uses the context received from `OnDecodedEvent` instead of embedding one at construct time.
- Add `OnExportEvent` hook system to the exporter to provide extension capabilities.
- Allow providing a custom writer to exporters.
- Allow providing a custom encoder to exporters.

Related: #35514

devodev committed:
daemon: extract hubble exporter logic into hive cell

Signed-off-by: Alexandre Barone <abalexandrebarone@gmail.com>

devodev committed:
hubble: use context from OnDecodedEvent in static exporter

Signed-off-by: Alexandre Barone <abalexandrebarone@gmail.com>

devodev commented on pkg/hubble/cell/cell.go:
```diff
@@ -36,6 +37,9 @@ var Cell = cell.Module(
 	cell.Config(defaultConfig),
+
+	// Provides Hubble file exporters.
+	exporter.Cell,
```
I am unsure if the hubble cell or the daemon cell should provide this?

devodev replied:
I think it makes more sense like it is now with hubble declaring the exporters cell.

devodev commented on pkg/hubble/exporter/exporter.go:
```diff
@@ -34,10 +34,7 @@ type exporter struct {
-	ctx context.Context,
-	logger logrus.FieldLogger,
-	options ...exporteroption.Option) (*exporter, error) {
+func NewExporter(logger logrus.FieldLogger, options ...exporteroption.Option) (*exporter, error) {
```
Having to provide a context at init time does not seem compatible with hive and how dependencies are handled. I could be wrong, please tell me.

In this case, I felt the context from OnDecodedEvent, which is used in the function, was probably more relevant than the one passed to hubble.launch ? Let me know if I got this wrong and there is more to that story.

devodev committed:
hubble: provide hubble exporters using observer options

Signed-off-by: Alexandre Barone <abalexandrebarone@gmail.com>

kaworu commented on pkg/hubble/exporter/cell.go:
```diff
@@ -0,0 +1,161 @@
+// Cell provides OnDecodeEvent handlers that export hubble events to files.
+var Cell = cell.Module(
+	"hubble-exporters",
+	"Hubble Exporters",
```
nit:
```suggestion
	"Hubble flow log exporters",
```

kaworu commented on pkg/hubble/exporter/cell.go:
```diff
@@ -0,0 +1,161 @@
+	"github.com/cilium/cilium/pkg/hubble/observer/observeroption"
+)
+
+// Cell provides OnDecodeEvent handlers that export hubble events to files.
```
nitpicking:
```suggestion
// Cell provides OnDecodeEvent handlers exporting hubble events to files.
```

kaworu commented on pkg/hubble/exporter/cell.go:
```diff
@@ -0,0 +1,161 @@
+func newHubbleStaticExporter(params hubbleExportersParams) (hubbleExportersOut, error) {
+	if err := params.Config.validate(); err != nil {
+		return hubbleExportersOut{}, fmt.Errorf("hubble-exporter configuration error: %w", err)
+	}
```
Both `newHubbleStaticExporter()` (this func) and `newHubbleDynamicExporter()` below calls  `params.Config.validate()`, so I guess we're running validation twice. Questions:
1. on error, do we get two similar / identical error log messages?
2. is there a way to call it only once?

devodev replied:
Found a clever way to do that by providing an intermediate `validatedConfig` type. Much nicer than validating twice indeed.
https://github.com/cilium/cilium/pull/35596/commits/8251918c3c3fdf9d568ef2baf9e7b31125c0eec3

kaworu commented on pkg/hubble/exporter/cell.go:
```diff
@@ -0,0 +1,161 @@
+		return hubbleExportersOut{}, fmt.Errorf("hubble-exporter configuration error: %w", err)
+	}
+
+	if params.Config.ExportFilePath != "" {
```
nit: I'd slightly prefer
```go
if !params.Config.ExportFilePath != "" {
	return hubbleExportersOut{}, nil
}
…
```
as it would reduce nesting a bit.

kaworu commented on pkg/hubble/exporter/cell.go:
```diff
@@ -0,0 +1,161 @@
+			exporterOpts = append(exporterOpts, exporteroption.WithOnExportEvent(onExportEvent))
+		}
+		exporter, err := NewExporter(params.Logger, exporterOpts...)
+		if err == nil {
```
again here shortcuting would help readers IMO
```go
		if err != nil {
			params.Logger.WithError(err).Error("Failed to configure Hubble static exporter")
			return hubbleExportersOut{}, nil
		}
		…
```
Also question: why don't we return the error to the caller?

devodev replied:
The current logic is non-failing at the moment and I wasn't sure if we wanted to change that behaviour in this PR.
But happy to do so if warrant this a fatal failure for the daemon.

chancez replied:
Is it possible for our cell to be marked non-critical so we can just return errors as normal, but not cause the entire agent to fail? Ideally this would be surfaced so that the "hubble flow export cell is unhealthy", which I know we have some mechanisms for, but I'm not sure exactly what they are.

devodev replied:
It seems possible to do something like that using job.OneShot for which the callback is passed a cell.Health we could use for that purpose, will look into it.

devodev replied:
By moving to job.OneShot for the dynamic exporter, we automatically get a health status attached, and it becomes degraded if an error is returned.

For the static exporter, since its hook-based, I didn't feel like it would make sense to declare a OneShot job only to declare a status (which is not logged btw and is only viewable using cilium-dbg).

kaworu commented on pkg/hubble/exporter/exporter.go:
```diff
@@ -139,6 +140,20 @@ func (e *exporter) OnDecodedEvent(ctx context.Context, ev *v1.Event) (bool, erro
+	for _, f := range e.opts.OnExportEvent {
+		stop, err := f.OnExportEvent(ctx, ev, e.encoder)
+		if err != nil {
+			e.logger.WithError(err).WithField("event", ev).Info("failed in OnExportEvent")
```
Generally, we don't want to log the full event above the Debug log level. First and foremost to avoid overloading logging but also there is sometimes a privacy concern (here I think it's fine because the event was meant to be exported anyway, and thus users should already be cautious wrt sensitive data, i.e. configure hubble redaction accordingly). cc @chancez what do you think?

devodev replied:
Good points. I actually replicated the same hook processing logic from `LocalObserverServer.Start()`. Happy to make changes / align them if we feel like doing that.

chancez replied:
I agree with Alex, and this should be an `Error` or `Warn` instead of `Info`. Additionally the log could probably phrased `OnExportEvent failed`.

devodev replied:
So remove the event field completely or log one/subset of fields? And should we also update `LocalObserverServer.Start()` to reflect that as well ?

devodev replied:
removed event logging and updated to Warn level

kaworu commented on pkg/hubble/exporter/exporter.go:
```diff
@@ -139,6 +140,20 @@ func (e *exporter) OnDecodedEvent(ctx context.Context, ev *v1.Event) (bool, erro
+		if stop {
+			// abort exporter pipeline by returning early but do not prevent
+			// other OnDecodedEvent hooks from firing
+			return false, nil
```
:+1:

kaworu reviewed (changes requested):
Nice work @devodev :+1: 

Some comments and questions inline. Also I noticed that in [`exporter.Stop()`](https://github.com/cilium/cilium/blob/51b5d7fbdfca7b434e4e3f0bd7ad221a2ebe2262/pkg/hubble/exporter/exporter.go#L125-L133) we're closing the writer, which can be setup with [`os.Stdout`](https://github.com/cilium/cilium/blob/51b5d7fbdfca7b434e4e3f0bd7ad221a2ebe2262/pkg/hubble/exporter/exporter.go#L51) and I don't think we're supposed to close stdout. Not to be addressed in this PR, but something to look into. cc @chancez whom I added as a reviewer as well since he has context about export.

glrf commented on pkg/hubble/exporter/exporter.go:
```diff
@@ -134,15 +131,29 @@ func (e *exporter) Stop() error {
 	select {
-	case <-e.ctx.Done():
+	case <-ctx.Done():
 		return false, e.Stop()
```
This doesn't seem correct.  You don't want to stop the exporter just because one `OnDecodedEvent` call was canceled.

devodev replied:
Ha good catch, and I know what happened. The exporter was previously using a stored context received from hubble.launch() instead of the passed in one. Fix incoming, thanks :)

devodev replied:
Fixed for both exporters in: https://github.com/cilium/cilium/pull/35596/commits/8933ccbb7213db54f3be5ee68868b3c24e1f42ed

glrf commented on pkg/hubble/exporter/exporter.go:
```diff
@@ -134,15 +131,29 @@ func (e *exporter) Stop() error {
 		return false, nil
 	}
+
+	// Process OnExportEvent hooks
```
The code seem fairly trivial, but could we still add a basic test case to check that we call all the registered hooks and that `stop` is handled correctly? Just to make sure that we don't break this going forward.

devodev replied:
Very good point, on it.

devodev replied:
Added a new test in: https://github.com/cilium/cilium/pull/35596/commits/81255abe387a19437e520748cfc179a789e6b548#diff-2fbc37e5ceace9551caf45e3067a7ac55b1d523c1f9392e3a0afba0e71203b08R274

glrf commented on pkg/hubble/exporter/cell.go:
```diff
@@ -0,0 +1,161 @@
+		return hubbleExportersOut{}, fmt.Errorf("hubble-exporter configuration error: %w", err)
+	}
+
+	if params.Config.FlowlogsConfigFilePath != "" {
```
Nit: Similarly to Alex's suggestion, IMO an early return here would be more readable

devodev committed:
hubble: introduce validatedConfig wrapper for exporter cell

Prevents having to call config validation in both hive providers.
Also refactor to avoid nested statements and update exporter
cell name.

Signed-off-by: Alexandre Barone <abalexandrebarone@gmail.com>

chancez commented on pkg/hubble/exporter/cell.go:
```diff
@@ -0,0 +1,161 @@
+		exporter := NewDynamicExporter(params.Logger, params.Config.FlowlogsConfigFilePath, params.Config.ExportFileMaxSizeMB, params.Config.ExportFileMaxBackups)
+		params.Lifecycle.Append(cell.Hook{
+			OnStart: func(hc cell.HookContext) error {
+				go exporter.Watch()
```
Let's add a TODO here to use the hive mechanism for async tasks instead of launching the Goroutine ourselves.

devodev replied:
I was actually going to ask, because it seems other hive cells are mixed between doing this and using job.OneShot. I assume OneShot is better for long-lived task whereas OnStart/OnStop is essentially doing point-in-time work that can block ?

With the recent refactor, we can do both, so let me update it.

devodev replied:
Updated to use job.OneShot, much nicer thanks.

chancez commented on pkg/hubble/exporter/config_watcher.go:
```diff
@@ -23,43 +22,40 @@ var reloadInterval = 5 * time.Second
+func (c *configWatcher) watch() {
 	// TODO replace ticker reloads with inotify watchers
-	watcher.ticker = time.NewTicker(reloadInterval)
-	watcher.stop = make(chan bool)
```
I'm a fan of the previous approach that allocates the ticker/stop in the `watch` method because it makes it possible to stop and start the watcher, which IMO simplifies things by not having to worry about if the object can be re-used after stopped.

devodev replied:
The thing is, it could never be restarted before because we didn't have a start method or a way to re-enable the ticker. I thought by doing it this way, we are more explicit about being single-use. I can always update start so it is compatible and can be restarted if desired.

devodev replied:
Updated to be context-based, similar to other cells making use of OneShot, which means we don't need to store any state afterall. It also has the advantage of decoupling starting/stoping the watcher from stopping the exporter itself (which is not restartable).

glrf commented on pkg/hubble/exporter/exporter.go:
```diff
@@ -132,17 +130,34 @@ func (e *exporter) Stop() error {
+	case <-ctx.Done():
+		return false, nil
 	default:
 	}
```
Nitpick: Now it's correct, but questionably useful. I don't think we need to handle the case where someone calls us with an already canceled context. (But it's also fine to keep it)

devodev replied:
Its a good point.

I guess it depends if we want best-effort early exit to be handled by each hook (and let them be the judge) or at the top level call site where it could be done systematically, and optionally by the hooks for long-lived/blocking logic.

devodev replied:
For this PR, I'll try and avoid behaviour changes and in a subsequent PR, we can look at cleaning this up.

glrf reviewed (approved):
LGTM :rocket:

bimmlerd commented on pkg/hubble/exporter/cell.go:
```diff
@@ -0,0 +1,170 @@
+type hubbleExportersOut struct {
+	cell.Out
+
+	ObserverOptions []observeroption.Option `group:"hubble-observer-options,flatten"`
```
I think these are too "small" to be in the object graph individually. In my head, hive is most useful at the component level - tying together a large application with many components. The heuristic for me is "Do I want to see XYZ in the `cilium hive` output?". With individual options for a component, I'd lean towards no. (The best docs for hive and its use in cilium is likely https://docs.cilium.io/en/latest/contributing/development/hive/#id1)

Not sure it needs to be part of this PR, and I wouldn't want to block this work on it, but you could potentially wrap these up into a single config/options struct with a slice and then `ProvidePrivate` it from the toplevel hubble module to just the hubble components? It depends on who needs to access this, but I'm guessing it's not the whole, global object graph.

devodev replied:
> hi, sorry for the drive-by review ✌️
>
> Maybe it's worth requesting a review from the foundations team when adding new cells :)

Thank you for the review, its very much appreciated, and agreed, will do that going forward!

If I understand correctly, your suggestion is to not create a Cell for the hubble exporters and instead provide these newX constructors directly on the Hubble Cell (and maybe as only one constructor)? Can we also declare multiple configuration as part of one cell? I also assumed that a Cell declared into a parent Cell is only available to parent and child, but I guess that is wrong?

Going forward, the Hubble cell will be broken up into 10-13 logical components. Wouldn't it make more sense for these to be Cells instead of plain providers for organization/config declaration purposes? The main idea behind this work is to break up this gigantic Hubble cell to make it easier to work with and isolate/declare dependencies. I feel if we declare everything on one Cell with one single configuration/metrics/etc, we would not gain much from the current setup and it would remain somewhat messy? But I might be misunderstanding your comment.

One thing that might not be clear is we are providing these exporters as options because we want to hook ourselves into the hubble observer service as an `OnDecodedEvent` hook, but these are fully fledge components, and they could also be provided as `FlowExporter`. It depends if we want Hubble to be aware of "exporters" or only of the hooks.

Happy to discuss further our use-case and find the best approach. I'll also CC @kaworu to get his input on this.

Thanks again for jumping in and helping us 😸

bimmlerd replied:
Hey, sorry for the delay

> Can we also declare multiple configuration as part of one cell?

Yeah, I wouldn't know why this would not work. `pkg/clustermesh/cell.go` does it too, for example.

> I also assumed that a Cell declared into a parent Cell is only available to parent and child, but I guess that is wrong?

If I understand correctly, yes, that's not true. Declaring a cell doesn't do much, but `Provide` is the key. What you `Provide` to the object graph is global, unless you use `ProvidePrivate`. Think of `Provide` as the way to declare the interface/service you're offering to the whole application. (To ensure I wasn't spreading misinformation I just edited the example in github.com/cilium/hive, `provide`d an `int` from a submodule and `invoke`d a `func(int)` top-level - it gets the `int`.)

I don't have enough insight in this area of the codebase to give well-founded answers to your questions, I'm afraid. I don't have a strong opinion on whether hubble should be a single cell or a few, as long as each cell either has a sensible API to offer to the rest of the application, or is private to only the hubble module.

If you run `go run ./daemon hive` you can look at some of the existing modules/cells, and see which parts of the are private (the :lock: emoji) to see some examples.

devodev replied:
Thanks for the details, I think I understand better what you mean and I agree that its good to keep dependencies encapsulated when possible.
I went ahead and pushed a new commit that removes the Hubble-exporters cell and instead ProvidePrivate what we need directly on the hubble cell.

bimmlerd reviewed (commented):
hi, sorry for the drive-by review :v: 

Maybe it's worth requesting a review from the foundations team when adding new cells :)

devodev committed:
hubble: replace exporter cell with cell.ProvidePrivate within hubble cell

Avoids providing hubble-specific dependencies globally.

Signed-off-by: Alexandre Barone <abalexandrebarone@gmail.com>

devodev committed:
hubble: allow providing custom encoder in flow exporter

Signed-off-by: Alexandre Barone <abalexandrebarone@gmail.com>

devodev committed:
hubble: allow providing custom writer in flow exporter

Decouple the exporter from its writer implementation.

Signed-off-by: Alexandre Barone <abalexandrebarone@gmail.com>

devodev committed:
hubble: merge exporteroption package into exporter

Signed-off-by: Alexandre Barone <abalexandrebarone@gmail.com>

chancez reviewed (approved):
Latest changes LGTM. I really like being able to configure the encoder so we can introduce different exporter output formats in the future.

glrf commented on pkg/hubble/exporter/exporter.go:
```diff
@@ -132,36 +137,37 @@ func (e *exporter) Stop() error {
-	case <-e.ctx.Done():
-		return false, e.Stop()
+	case <-ctx.Done():
+		return false, nil
```
Nit: I think we should return a `ctx.Err()`

glrf commented on pkg/hubble/exporter/dynamic_exporter.go:
```diff
@@ -10,30 +10,60 @@ import (
 	select {
 	case <-ctx.Done():
-		return false, d.Stop()
+		return false, nil
```
Nit: This should probably also return `ctx.Err()`

glrf commented on pkg/hubble/exporter/cell/cell.go:
Nitpick: This is a bit of a pet peeve of me, but do we really need this separate package? Couldn't this live in `pkg/hubble/exporter`?

devodev replied:
It could yess. I've tried to follow what we did for hubble with hubblecell, as well as other cells with this pattern. I kinda liked that the cell was forced to declare its own API and rely on exported types for its underlying implementation. Keep sthings neatly decoupled.

But both work, and both are used in the hive right now, I don't have a strong opinion.

glrf replied:
Alright. I think it's fine to keep it. It's just personal preference

glrf reviewed (commented):
Overall this looks good to me. It's definitely an improvement and I'm not going to block this from being merged.

However, I think I agree with David's point of `observeroption.Option` being too small and generic for the hive object graph. Currently no cell provides this to the global object graph, but the Hubble Cell still takes it as an input. This means that any Cell can still provide arbitrary options to the Hubble cell, which feels weird. An observer option is not really a service a cell can `Provide`.

In my mind it would for example make more sense if we'd basically provide the exporter itself. Maybe something like

```
// Provided by exporter cell
type Exporter interface {
  Export(ctx context.Context, ev *v1.Event) (bool, error)
}

...

// In hubble cell
for _, e := range h.exporters {
  observerOpts = appned(observerOpts, observeroption.WithOnDecodedEvent(e.Export)
}
```

It's very nitpicky, but `Export` is the service the exporter actually provides not the observer option. Also that way we don't open the hubble cell to receiving arbitrary observer options from arbitrary cells, which IMO can lead to fairly weird and unexpected spooky actions at a distance and harder to follow dependencies.

devodev committed:
hubble: add OnExportEvent hook system in exporter

Signed-off-by: Alexandre Barone <abalexandrebarone@gmail.com>

devodev committed:
hubble: provide exporters using the FlowLogExporter interface

Signed-off-by: Alexandre Barone <abalexandrebarone@gmail.com>

glrf reviewed (approved):
Nice work :rocket:

pippolo84 reviewed (approved):
Focused my review on the newly added cells and the usage of the "hive and cells" framework. LGTM, thanks! :100:

kaworu reviewed (approved):
Awesome work @devodev :clap: !
